interfaces/prompting/patterns/patterns.go

```diff
@@ -0,0 +1,175 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package patterns
+
+import (
+	"encoding/json"
+	"fmt"
+	"regexp"
+	"strings"
+
+	doublestar "github.com/bmatcuk/doublestar/v4"
+)
+
+// Limit the number of expanded path patterns for a particular pattern.
+// When fully expanded, the number of patterns for a given unexpanded pattern
+// may not exceed this limit.
+const maxExpandedPatterns = 1000
+
+// PathPattern is an iterator which yields expanded path patterns.
+type PathPattern struct {
+	original   string
+	renderTree renderNode
+}
+
+// ParsePathPattern validates the given pattern and parses it into a PathPattern
+// from which expanded path patterns can be iterated, and returns it.
+func ParsePathPattern(pattern string) (*PathPattern, error) {
+	pathPattern := &PathPattern{}
+	if err := pathPattern.parse(pattern); err != nil {
+		return nil, err
```

Comment:
```suggestion
		return nil, fmt.Errorf("cannot parse path pattern %q: %w", pattern, err)
```
end even then I would question whether you need this prefix at all, since the caller knows the pattern already

Reply:
I want the `parse` method to handle its error messages rather than `ParsePathPattern`, since `parse` is called by `UnmarshalJSON` and I think the error should be properly prefixed there as well.